bin/all_sky_search/pycbc_template_kde_calc

```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python
+
+import numpy, h5py, operator, argparse, logging
+from pycbc import init_logging
+import pycbc.conversions as convert
+from pycbc import libutils
+akde = libutils.import_optional('awkde')
+kf = libutils.import_optional('sklearn.model_selection')
+
+parser = argparse.ArgumentParser(description=__doc__)
+parser.add_argument('--signal-file')
+parser.add_argument('--template-file', required=True)
+parser.add_argument('--output-file', required=True)
+parser.add_argument('--make-signal-kde', action='store_true')
+parser.add_argument('--make-template-kde', action='store_true')
+parser.add_argument('--verbose', action='count')
+args = parser.parse_args()
+init_logging(args.verbose)
+
+def kde_awkde(x, x_grid, alp=0.5, gl_bandwidth='silverman', ret_kde=False):
+    kde = akde.GaussianKDE(glob_bw=gl_bandwidth, alpha=alp, diag_cov=True)
+    kde.fit(x)
+    if isinstance(x_grid, (list, tuple, numpy.ndarray)) == False:
+        y = kde.predict(x_grid)
+    else:
+        y = kde.predict(x_grid)
+    if ret_kde == True:
+        return kde, y
+    return y
+
+def kfcv_awkde(sample, bwchoice, alphachoice, k=2):
+    if bwchoice not in ['silverman', 'scott']:
+        bwchoice = float(bwchoice)
+    fomlist = []
+    kfold = kf.KFold(n_splits=k, shuffle=True, random_state=None )
+    for train_index, test_index in kfold.split(sample):
+        train, test = sample[train_index], sample[test_index]
+        y = kde_awkde(train, test, alp=alphachoice, gl_bandwidth=bwchoice)
+        fomlist.append(numpy.sum(numpy.log(y)))
+    return numpy.mean(fomlist)
+
+def optimizedparam(sampleval, nfold=2):
+    bwgrid = ['scott', 'silverman']+numpy.logspace(-2,0,10).tolist()
+    alphagrid = [1]
+    FOM= {}
+    for gbw in bwgrid:
+        for alphavals in alphagrid:
+            FOM[(gbw, alphavals)] = kfcv_awkde(sampleval, gbw, alphavals, k=nfold)
+    optval = max(FOM.items(), key=operator.itemgetter(1))[0]
+    optbw, optalpha  = optval[0], optval[1]
+    maxFOM = FOM[(optbw, optalpha)]
+    return optbw, optalpha
+
+#calling template data and constructing template_kde_hdf file over template params
+template_file = h5py.File(args.template_file, 'r')
+mass1_n = template_file['mass1'][:]
+mass2_n = template_file['mass2'][:]
+spin1z_n = template_file['spin1z'][:]
+spin2z_n = template_file['spin2z'][:]
+chi_eff_n = convert.chi_eff(mass1_n, mass2_n, spin1z_n, spin2z_n)
+mchirp_n = convert.mchirp_from_mass1_mass2(mass1_n, mass2_n)
+eta_n = convert.eta_from_mass1_mass2(mass1_n, mass2_n)
+
+if args.make_template_kde:
+    sample_n = numpy.vstack((numpy.log(mchirp_n), eta_n, chi_eff_n)).T
+    grid_pts_n = sample_n
+    optbw_n, optalpha_n = optimizedparam(sample_n, nfold=2)
```

Comment:
I'd suggest to include a logging info statement just above this line here, with a message such as 'Starting optimization of template KDE parameters ..'